[pt1][quant] Dynamic Quantized Linear module

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23139 [pt1][quant] Unit Test for the Dynamic Quantized Linear operator
* #24231 [WIP][pt1][quant] Add the default_weight_observer for the dynamic quantization path
* #24299 [pt1][quant] Remove the activation observer for default_qconfig
* #24232 [pt1][quant] Make the default qconfig_dict
* #23330 [pt1][quant] Fix the dimension mismatch issues when running the BERT model
* #23212 [pt1][quant] Add the type matching rule for qconfig_dict
* **#23128 [pt1][quant] Dynamic Quantized Linear module**
* #23464 [pt1][quant] Add dynamic quantized Linear op in PyTorch

- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
[pt1][quant] Dynamic Quantized Linear operator and module

- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jerryzh168 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+import torch
+from ....modules.module import Module
```
not used?

jianyuh replied:
Will update.

jianyuh replied:
Updated.

jerryzh168 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+from ....modules.module import Module
+from ....modules.linear import Linear as NNLinear
+
+class Linear(NNLinear):
```
you can inherit from nnq.Linear if you want to reuse `.from_float`

jianyuh replied:
Then we need to add some logic to determine whether we should add the activation observer?

raghuramank100 replied:
Yes, please inherit from nnq.Linear. This way, you inherit loading/saving of state dict.  I would recommend the following changes:

```python
class Linear(nn.quantized.Linear):
  
    __constants__ = ['bias', 'in_features', 'out_features']
    def __init__(self, in_features, out_features, bias=True):
        super(Linear, self).__init__(in_features, out_features, bias)
           if bias:
             del self.bias
             bias= torch.Tensor(out_features)
             self.register_buffer('bias', bias)
        else:
            self.register_buffer('bias', None)
      del self.scale
      del self.zero_point

    def forward(self, x):
        # Note that we can handle self.bias == None case.
        Y = torch.ops.quantized.fbgemm_linear_dynamic(
            x, self._packed_weight,
            self.bias)
        return Y.to(x.dtype)

 @staticmethod
    def from_float(mod):
        r"""Create a quantized module from a float module or qparams_dict
            Args: `mod` a float module, either produced by torch.quantization utilities
            or directly from user
        """
   
            assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
            assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
            weight_observer = mod.qconfig.weight()
            weight_observer(mod.weight)
           wt_scale, wt_zp = weight_observer.calculate_qparams()
           qweight = torch.quantize_linear(mod.weight.float(), wt_scale, wt_zp.long().item(), torch.qint8)
           qlinear = Linear(mod.in_features, mod.out_features, mod.bias) # Instantiate dynamic.linear module
           qlinear._packed_weight = torch.ops.quantized.fbgemm_linear_prepack(qweight)
         # Can handle bias being None by just copying it to qLinear.bias
          qlinear.bias = mod.bias
     
         return qlinear

dzhulgakov replied:
I'm not sure inheriting and killing so many attributes is worth it. Maybe Static QLinear should inherit from this one instead? (or have common base class).

I'm also fine with duplication of getstate/setstate. We likely can simplify it further once custom C++ classes are fully ready.

jianyuh replied:
I currently adopted the approach of inheriting from `nnq.Linear`, as this one is more straightforward. Of course a better way is to have the common base class.

raghuramank100 replied:
Please land this version first and create a follow up task for refactoring Linear to have a common base class.

jianyuh replied:
Filed a task.

jerryzh168 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+
+    # TODO: support initializing from quantization parameters when Quantizer is
+    # exposed in python
+    @staticmethod
```
also we use classmethod now.

jerryzh168 replied:
this is not updated yet, but I'll probably update that.

jianyuh replied:
Updated.

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -220,6 +226,16 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
+    """
+    model.eval()
+    model = prepare_dynamic(model, qconfig_dict)
+    if run_fn is not None:
```
I think we can remove `run_fn` and `run_args`?

jianyuh replied:
Right: we don't need the evaluation/calibration for the dynamic quantization path.

jianyuh replied:
Updated.

jerryzh168 commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+        torch.Size([128, 30])
+    """
+
+    __constants__ = ['bias', 'in_features', 'out_features', 'scale', 'zero_point']
```
do we need these? is this used in script module?

jianyuh replied:
Right: this is used in the script module. According to https://pytorch.org/docs/stable/jit.html:
> Attributes of a ScriptModule can be marked constant by listing them as a member of the __constants__ property of the class

I just followed nnq.Linear. Should that part also be updated?

jerryzh168 replied:
since they are not ScriptModule, I think we can remove them

jianyuh replied:
Will update that.

jianyuh replied:
Removed the `__constants__`.

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -220,6 +226,16 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
     convert(model)
     return model
 
+def quantize_dynamic(model, run_fn=None, run_args=None, qconfig_dict=None):
```
em... why do we even need 'run' function for dynamic quantization?! the only observers you need are the weight observers and those don't need any training data

jianyuh replied:
Right. Will update it and remove the run_fn and run_args arguments.

jianyuh replied:
Updated.

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -220,6 +226,16 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
+    r"""Converts a float model to dynamic quantized model. Do dynamic training and output a quantized model.
+    """
+    model.eval()
+    model = prepare_dynamic(model, qconfig_dict)
```
we should make default qconfig_dict or something. Basically if I call `quantize_dynamic(model)` it should just go and swap out all known layers with default observer settings.

raghuramank100 replied:
Agreed, can we set default qconfig_dict to correspond to nn.Linear (after your change https://github.com/pytorch/pytorch/pull/23212)

jianyuh replied:
Will submit a PR on top of https://github.com/pytorch/pytorch/pull/23212 for this change.

jianyuh replied:
Added the default qconfig_dict in https://github.com/pytorch/pytorch/pull/24232.

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -220,6 +226,16 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
+    model = prepare_dynamic(model, qconfig_dict)
+    if run_fn is not None:
+        run_fn(model, run_args)
+    convert(model, DEFAULT_DYNAMIC_MODULE_MAPPING)
```
pass DEFAULT_DYNAMIC_MODULE_MAPPING as arguments too

jianyuh replied:
Will submit a PR on top of #23212 for this change.

jianyuh replied:
DEFAULT_DYNAMIC_MODULE_MAPPING is now the default value for an argument `mapping` in the `quantize_dynamic` function.

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -260,6 +280,9 @@ def convert(module, mapping=DEFAULT_MODULE_MAPPING):
 
     return module_swapped
 
+def convert_dynamic(module, mapping=DEFAULT_DYNAMIC_MODULE_MAPPING):
```
I don't think we need to expose it. Since majority of folks would be calling just `quantize_dynamic` (which sets observers), there's less need in this toplevel API. (note that in PTQ or QAT there's a higher need in separate apis as user needs to insert training/eval loop in the middle)

raghuramank100 replied:
Similarly, we do not need to expose prepare_dynamic as well. Quantize_dynamic is sufficient.

jianyuh replied:
Right: this is mainly for the unit test purposes, e.g., https://github.com/pytorch/pytorch/pull/23128/files/66528008d7b9bca74fa6817b77d9863d1ce2bb1d#diff-aceee327fb8c71ff9723e11597f7cbc7R298-R303. In the unit test, we need the `prepare_dynamic` and `convert_dynamic` for verifying the function correctness. Maybe we should just re-use `prepare_dynamic` and `convert`, and add an argument for `use_dynamic` to specially handle the dynamic quant cases?

jerryzh168 replied:
we can put these functions in common_quantization.py

jianyuh replied:
Put these functions in `common_quantization.py`, only for the purpose of the unit test.

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+from ....modules.module import Module
+from ....modules.linear import Linear as NNLinear
+
+class Linear(NNLinear):
```
can you please add setstate/getstate like other modules to make sure this module is serializable?

also please verify that it works with JIT (not sure whether weakscript annotation propagates to child classes)

jianyuh replied:
Will add the setstate/getstate like other modules for the serialization

jianyuh replied:
I currently adopted the approach of inheriting from `nnq.Linear`, so we now can just reuse the setstate/getstate from the base class. Of course a better way is to have the common base class. Will file a task to verify it works with JIT and check whether weakscript annotation propagates to child classes. For PyTorch 1.2 release, we currently only do the eager mode for the Linear module (this is not Script Module yet).

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+        super(Linear, self).__init__(in_features, out_features, bias)
+        del self.weight
+        del self.bias
+        bias_fp32 = torch.Tensor(out_features).float()
```
nit: no need in .float

jianyuh replied:
Will update it.

jianyuh replied:
Updated.

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+
+    __constants__ = ['bias', 'in_features', 'out_features', 'scale', 'zero_point']
+
+    def __init__(self, in_features, out_features, bias=True):
```
seems like you ignore bias=False/True

jianyuh replied:
Right: Will add the handler for bias = None case.

jianyuh replied:
Updated.

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,76 @@
+        qweight = torch.quantize_linear(mod.weight.float(), wt_qparams[0], wt_qparams[1].long().item(), torch.qint8)
+        qlinear = Linear(mod.in_features, mod.out_features)
+        qlinear._packed_weight = torch.ops.quantized.fbgemm_linear_prepack(qweight)
+        qlinear.bias = mod.bias.float()
```
input module might not have bias at all

jianyuh replied:
You are right. This line reports the error when I dynamic quantize the BERT (some Linear modules have bias = None). I updated the OSS part locally but forget to submit the changes. Sorry for rushing to enable the dynamic quantization. Will update this PR tomorrow.

jianyuh replied:
Updated.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -260,6 +260,166 @@ def checkQuantized(model):
+        # Check if observers are inserted
+        self.checkObservers(model)
+
+        test_only_eval_fn(model, self.calib_data)
```
Can we remove this line? We do not need it for dynamic quant.

jianyuh replied:
Will update it.

jianyuh replied:
Updated.

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -150,6 +151,11 @@ def prepare_qat(model, qconfig_dict=None):
 
+def prepare_dynamic(model, qconfig_dict=None):
+    propagate_qconfig(model, qconfig_dict)
+    add_observer(model)
```
Remove add_observer, we do not need it for dynamic quant. We can address the dependency in swap_module by modifying it as follows:
```python
def swap_module(mod, mapping):
    r"""Swaps the module if it has a quantized counterpart and it has an
    `observer` attached.
    Args:
        mod: input module
        mapping: a dictionary that maps from nn module to nnq module
    Return:
        The corresponding quantized module of `mod`
    """
    new_mod = mod
    if hasattr(mod, 'qconfig'): # Instead of observer, use qconfig
        if type(mod) in mapping:
            new_mod = mapping[type(mod)].from_float(mod)

    return new_mod
```

jerryzh168 replied:
no we can't, qconfig can be None. we are using observer to mark whether we want quantization or not right now.

jianyuh replied:
As @jerryzh168 mentioned, there will be an issue here.

raghuramank100 replied:
@jerryzh168 and myself discussed this, he is going to land a diff that removes the check for observers in swap module (basically the code below):
So, now you can remove observers as a requirement for dynamic quantization.

def swap_module(mod, mapping):
r"""Swaps the module if it has a quantized counterpart and it has an
    `observer` attached.
    Args:
        mod: input module
        mapping: a dictionary that maps from nn module to nnq module
    Return:
        The corresponding quantized module of `mod`
    """
    new_mod = mod
    if hasattr(mod, 'qconfig') and mod.qconfig is not None: # Instead of observer, use qconfig
        if type(mod) in mapping:
            new_mod = mapping[type(mod)].from_float(mod)

    return new_mod

jianyuh replied:
Got it! Will modify this.

jianyuh replied:
This has been done by @jerryzh168 in https://github.com/pytorch/pytorch/pull/23561. Updated the PR by removing the `add_observer`.

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -220,6 +226,16 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
     convert(model)
     return model
 
+def quantize_dynamic(model, run_fn=None, run_args=None, qconfig_dict=None):
```
Lets ensure that all three quantize functions take DEFAULT_MODULE_MAPPING as an argument, with a default value set. For QAT, we will need one more argument DEFAULT_QAT_MODULE_MAPPING. (cc @jerryzh168)

jerryzh168 replied:
yeah, we will expose this in top level quantize function.

jianyuh replied:
Exposed the `DEFAULT_MODULE_MAPPING` as an argument in this PR.

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

raghuramank100 commented on test/test_nn_quantized.py:
```diff
@@ -32,6 +33,26 @@ def test_relu_api(self):
+        Z_ref = torch.ops.quantized.fbgemm_linear_dynamic(X, W_pack_col_offset, B)
+
+        self.assertEqual(Z_ref, Z_dq)
+
```
A general test that we need to add for all nn.quantized modules is to ensure that scripting works.

jianyuh replied:
Good catch! I didn't update this unit test. Will update it.

jianyuh replied:
Updated.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -261,6 +261,166 @@ def checkQuantized(model):
+            self.checkDynamicQuantizedLinear(model.fc3)
+            self.checkDynamicQuantizedLinear(model.sub2.fc1)
+            self.checkLinear(model.sub2.fc2)
+            test_only_eval_fn(model, self.calib_data)
```
Please remove all calls to test_only_eval_fn: We do not need this for dynamic quant.

jianyuh replied:
Updated.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -261,6 +261,166 @@ def checkQuantized(model):
+        """
+        model = SingleLayerLinearModel().eval()
+        qconfig_dict = {
+            '': default_qconfig
```
Please add a default_qconfig for dynamic quant that does not have activation observers

jianyuh replied:
This doesn't work: the activation observers are the required positional argument for QConfig in https://github.com/pytorch/pytorch/blob/master/torch/quantization/QConfig.py#L6-L7.

raghuramank100 reviewed (changes requested):
Please see suggested changes

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,72 @@
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        assert type(mod) == NNLinear, 'nnq.Linear.from_float only works for nn.Linear'
```
nn.quantized.dynamic.Linear in this message?

jianyuh replied:
Fixed.

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,72 @@
+        # assert type(mod) == cls.__FLOAT_MODULE, ' nnq.' + cls.__name__ + '.from_float only works for ' + \
+        #                     cls.__FLOAT_MODULE.__name__
+        assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
+        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
```
you don't need observer, do you?

jianyuh replied:
Fixed.

dzhulgakov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,72 @@
+        #                     cls.__FLOAT_MODULE.__name__
+        assert hasattr(mod, 'qconfig'), 'Input float module must have qconfig defined'
+        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
+        weight_observer = mod.qconfig.weight()
```
given that dynamic quantization is so simple to use, I'd allow also to not have qconfig specified in which case you use the default one. Also, maybe also allow to override qconfig (or rather weight_observer) as an input argument?

jianyuh replied:
cc @jerryzh168 . After https://github.com/pytorch/pytorch/pull/23561, for dynamic quantization, we don't need to insert the observer.

dzhulgakov replied:
what I meant is to have the signature `def from_float(cls, mod, weight_observer=default_weight_observer)`. That way I don't even need to do prepare for the simple case - I can just pass in `nnd.Linear(mod)` like in https://github.com/pytorch/pytorch/blob/master/torch/jit/quantized.py#L13

jianyuh replied:
This is doable. However, it caused the circular import issues in Python. Basically, when we use `default_weight_observer` in `torch/nn/quantized/dynamic/modules/linear.py`, we need to import `torch.quantization` in `torch.quantized.dynamic` namespace. In the meanwhile, we also need to import `torch.quantized.dynamic` in `torch.quantization` as we need to convert the modules. The error message is like the following:
```
stderr:
Traceback (most recent call last):
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform007/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/fbcode/platform007/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 845, in <module>
    sys.exit(main(sys.argv))
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 841, in main
    return MainProgram(sys.argv).run()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 695, in run
    test_suite = self.load_tests()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 678, in load_tests
    suite = loader.load_all()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/__fb_test_main__.py", line 467, in load_all
    __import__(module_name, level=0)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/test_nn_quantized.py", line 8, in <module>
    import torch.nn.quantized.dynamic as nnqd
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/nn/quantized/dynamic/__init__.py", line 4, in <module>
    from .modules import *  # noqa: F401
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/nn/quantized/dynamic/modules/__init__.py", line 3, in <module>
    from .linear import Linear
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/nn/quantized/dynamic/modules/linear.py", line 5, in <module>
    from torch.quantization.observer import default_weight_observer
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/quantization/__init__.py", line 2, in <module>
    from .quantize import *  # noqa: F401
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantized#binary,link-tree/torch/quantization/quantize.py", line 4, in <module>
    import torch.nn.quantized.dynamic as nnqd
AttributeError: module 'torch.nn.quantized' has no attribute 'dynamic'
```
Working on fixing this. Let me know if you have any insights.

jianyuh replied:
Currently the observer class and qconfig class are in “torch.quantization” namespace. If we need the observer and qconfig in “torch.nn.quantize.dynamic”, we need to import “torch.quantization” (Check one example: https://github.com/pytorch/pytorch/pull/23128#discussion_r309528932). In the meanwhile, in the “torch.quantization” namespace, we need to import “torch.nn.quantize.dynamic”, because we need to convert the modules and specify the “torch.nn.quantize.dynamic.Linear” in the mapping. 

The possible solution is to move the “QConfig” and “observer” out of torch.quantization namespace.

Ping @dzhulgakov , @raghuramank100 , @jerryzh168 for suggestions.

dzhulgakov replied:
I think the standard recommendation is to move observer and qconfig into a separate submodule (hidden, just import everything in __init__ as well), then torch.nn.quantize.dynamic can import that submodule instead of the main one

jianyuh replied:
I would like to resolve in another PR, as I need to add the `default_weight_observer` for both `torch.quantized.dynamic.Linear` and `torch.quantized.Linear`. In addition, I need to move the observer and qconfig into a separate submodule.

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add a unit test for the Dynamic Quantized Linear operator (```torch.fbgemm_linear_quantize_weight```, ```torch.fbgemm_pack_quantized_matrix```, and ```torch.fbgemm_linear_int8_weight```) in ```test_quantized.py```.
- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```. ~~This is in a rudimentary stage. Will add more functions later~~.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -221,6 +222,15 @@ def quantize_qat(model, run_fn, run_args, qconfig_dict=None):
+def quantize_dynamic(model, qconfig_dict=None):
+    r"""Converts a float model to dynamic quantized model. Do dynamic training and output a quantized model.
+    """
+    model.eval()
```
why do you need .eval here?

raghuramank100 replied:
Both static and dynamic quant work only for models in eval(). This is to ensure that batch norm is handled correctly.

jianyuh replied:
Right: dynamic quantization only works for the inference.

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear operator and module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

dzhulgakov commented on test/common_quantization.py:
```diff
@@ -49,6 +51,12 @@ def test_only_train_fn(model, train_data, loss_fn=_default_loss_fn):
+    convert(module, DEFAULT_DYNAMIC_MODULE_MAPPING)
+
+def prepare_dynamic(model, qconfig_dict=None):
+    propagate_qconfig(model, qconfig_dict)
```
I guess qconfig_dict is already gone

jianyuh replied:
This is a utility function only for the testing purpose. Do you mean that we should remove the `qconfig_dict` in this function call (`propagate_qconfig` still has `qconfig_dict` argument: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L33)?

For dynamic quantization, we will still keep qconfig_dict (discussed with @raghuramank100 , @jerryzh168 ). Will submit a Diff to set default `qconfig_dict` to correspond to `torch.nn.quantized.dynamic.Linear` (after PR #23212), as you and @raghuramank100 suggested.

zafartahirov commented on test/common_quantization.py:
```diff
@@ -9,10 +9,12 @@
 from torch.quantization import QuantWrapper, QuantStub, DeQuantStub, \
     default_qconfig, QConfig, default_observer, default_weight_observer, \
-    default_qat_qconfig
+    default_qat_qconfig, propagate_qconfig, convert, DEFAULT_DYNAMIC_MODULE_MAPPING
```
We need to factor out all the mappings -- I will send a PR today. @jerryzh168

jianyuh replied:
Let me know if there is any update. Currently the mappings are still in this file.

zafartahirov commented on test/common_quantization.py:
```diff
@@ -104,6 +112,13 @@ def checkQuantizedLinear(self, mod):
         self.assertEqual(type(mod), nnq.Linear)
         self.assertEqual(mod.bias.dtype, torch.qint32)
 
+    def checkDynamicQuantizedLinear(self, mod):
```
Isn't this check too specific to be put in the common file? If you really need to factor something out in the test, but the method is too specific, you can always create a function in the child test class that you are creating, instead of the common parent class.

jerryzh168 replied:
this looks fine, it's checking on the leaf linear module.

zafartahirov commented on test/test_nn_quantized.py:
```diff
@@ -34,6 +36,67 @@ def test_relu_api(self):
+        out_features=st.integers(4, 8),
+        use_bias=st.booleans(),
+    )
+    def test_dynamic_linear_api(self, batch_size, in_features, out_features, use_bias):
```
If there will be a lot of dynamic quantization stuff, maybe we should create a separate test class. Maybe `DynamicQuantizationTest`?

jianyuh replied:
Will fix this.

jianyuh replied:
Updated.

zafartahirov commented on test/test_nn_quantized.py:
```diff
@@ -34,6 +36,67 @@ def test_relu_api(self):
+    )
+    def test_dynamic_linear_api(self, batch_size, in_features, out_features, use_bias):
+        """test API functionality for nn.quantized.dynamic.Linear"""
+        W = torch.rand(out_features, in_features).float()
```
FYI: I saw `rand` outside the `hypothesis` sometimes makes the tests "flaky". No need to change anything, but in case you notice falkiness, this might be causing it

zafartahirov commented on test/test_nn_quantized.py:
```diff
@@ -34,6 +36,67 @@ def test_relu_api(self):
+        self.assertEqual(model_dict['weight'], W_q)
+        if use_bias:
+            self.assertEqual(model_dict['bias'], B)
+        with tempfile.NamedTemporaryFile() as f:
```
if you are not going to use the file outside the `with` clause, it is better to use the `tempfile.TemporaryFile` instead -- that way, the file will be destroyed right after the `with` statement.

jianyuh replied:
Replace `tempfile.NamedTemporaryFile` with `tempfile.TemporaryFile` in this file.

zafartahirov commented on test/test_nn_quantized.py:
```diff
@@ -34,6 +36,67 @@ def test_relu_api(self):
+        Z_dq2 = qlinear(X)
+        self.assertEqual(Z_dq, Z_dq2)
+
+        # test serialization of module directly - will add this later
```
cleanup?

jianyuh replied:
Updated.

zafartahirov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,71 @@
+    We adopt the same interface as `torch.nn.Linear`, please see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to `torch.nn.Linear`, attributes will be randomly initialized at
```
``:class:`torch.nn.Linear` ``

jianyuh replied:
Updated.

zafartahirov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,71 @@
+            self.bias)
+        return Y.to(x.dtype)
+
+    # TODO: support initializing from quantization parameters when Quantizer is
```
This TODO might outlive us all. Instead of creating code debt, consider creating an issue or a task

jianyuh replied:
Seems this issue has been resolved, as the same `TODO` is removed in `torch/nn/quantized/modules/linear.py`.

zafartahirov commented on torch/nn/quantized/dynamic/modules/linear.py:
```diff
@@ -0,0 +1,71 @@
+        wt_qparams = weight_observer.calculate_qparams()
+        qweight = torch.quantize_linear(mod.weight.float(), wt_qparams[0], wt_qparams[1].long().item(), torch.qint8)
+        qlinear = Linear(mod.in_features, mod.out_features)
+        qlinear._packed_weight = torch.ops.quantized.fbgemm_linear_prepack(qweight)
```
We already have a property for the weights, you can use

```
qlinear.weight = qweight
```

Otherwise, this part of code might get outdated if we decide to change the semantics of the weight/_acked_weight

jamesr66a replied:
We need to revisit these properties. They're not supported in TorchScript

jianyuh replied:
Since you have made significant changes to the `torch.nn.quantized.Linear` module, I will adapt `torch.nn.quantized.dynamic.Linear` based on your changes.

jianyuh replied:
Updated based on @jamesr66a 's most recent changes.

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -198,14 +219,27 @@ def quantize(model, run_fn, run_args):
     return model
 
-def quantize_qat(model, run_fn, run_args):
+def quantize_qat(model, run_fn, run_args, mapping=DEFAULT_QAT_MODULE_MAPPING):
```
we actually have two mappings in qat, and this change seems unrelated, revert?

jianyuh replied:
Reverted.

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jianyuh committed:
Update on "[pt1][quant] Dynamic Quantized Linear module"


- Add the Dynamic Quantized Linear module in ```torch/nn/quantized/modules/linear.py```.
- Add a unit test for the Dynamic Quantized Linear module  in ```test_nn_quantized.py```.
- Add a unit test for the Model-level Quantization API.

Differential Revision: [D16258664](https://our.internmc.facebook.com/intern/diff/D16258664/)

jamesr66a commented on test/test_nn_quantized.py:
```diff
@@ -34,6 +36,78 @@ def test_relu_api(self):
+        # Simple round-trip test to ensure weight()/set_weight() API
+        self.assertEqual(qlinear.weight(), W_q)
+        W_pack = qlinear._packed_weight
+        qlinear._packed_weight = W_pack
```
What's the purpose of these two lines?

jianyuh replied:
Oops...This might be added during some rebase...Fixed in https://github.com/pytorch/pytorch/pull/24408.
